Update API pluralization docs (#5996)
kubernetes/community: contributors/devel/sig-architecture/api_changes.md

This new field must be inclusive of the singular field.  In order to satisfy
the compatibility rules you must handle all the cases of version skew, multiple
clients, and rollbacks.  This can be handled by admission control or API
registry logic (e.g. strategy) linking the fields together with context from
the API operation to get as close as possible to the user's intentions.

Upon any read operation:
  * If plural is not populated, API logic must populate plural as a one-element
    list, with plural[0] set to the singular value.

Upon any create operation:
  * If only the singular field is specified (e.g. an older client), API logic
    must populate plural as a one-element list, with plural[0] set to the
    singular value.  Rationale: It's an old client and they get compatible
    behavior.
  * If both the singular and plural fields are specified, API logic must
    validate that plural[0] matches the singular value.
  * Any other case is an error and must be rejected.  This includes the case of
    the plural field being specified and the singular not.  Rationale: It's
    impossible to tell the difference between an old client clearing the

>     singular field via patch and a new client setting the plural field.  For

Reviewer: so, let me see if I get it right, since this is they key of the change but I can't see it, is this the scenario you are describing here?
```
type Frobber struct {
  Height int           `json:"height"`
  Width  int           `json:"width"`
  Param  string        `json:"param"`  // the first param
  Params []string      `json:"params"` // all of the params
}
```

old client
```
PATCH  /frobbers/1
{ param: "" }
```
new client
```
PATCH  /frobbers/1
{ params: ["new"]}
```

Author: Not patch, but PUT.

Assume the user starts with:

```
kind: Frobber
height: 42
width: 3
param: "super"
```

On create we can set `params: ["super"]`.

On an unrelated POST (aka replace), an old client would send:

```
kind: Frobber
height: 3
width: 42
param: "super"
```

and a new client would send:

```
kind: Frobber
height: 3
width: 42
params: ["super"]
```

That seems clear enough - we can assume `param: "super"`.

But that same old client could send a this, via patch:

```
PATCH  /frobbers/1
{ param: "" }
```

That gets applied to the old object before registry code can see it, and we end up with:

```
kind: Frobber
height: 42
width: 3
params: ["super"]
```

By the previous logic, we would end up with `param: "super"`.  But that's a) not what the user wanted; and b) different than what happened before we pluralized.

To totally disambiguate that, we require users of `params` to always deal with `param`, too.

There may be other ways we could disambiguate, but we settled on this for clusterIP(s).

Reviewer: > clearing the singular field via patch and a new client setting the plural field.  For

heh, it is hard to me to get from that sentence all the implications, your comment is a really nice example I think is worth adding it

Author: By Your Command!  Pushed.

This new field must be inclusive of the singular field.  In order to satisfy
the compatibility rules you must handle all the cases of version skew, multiple
clients, and rollbacks.  This can be handled by admission control or API
registry logic (e.g. strategy) linking the fields together with context from
the API operation to get as close as possible to the user's intentions.

Upon any read operation:
  * If plural is not populated, API logic must populate plural as a one-element
    list, with plural[0] set to the singular value.

Upon any create operation:
  * If only the singular field is specified (e.g. an older client), API logic
    must populate plural as a one-element list, with plural[0] set to the
    singular value.  Rationale: It's an old client and they get compatible
    behavior.
  * If both the singular and plural fields are specified, API logic must
    validate that plural[0] matches the singular value.
  * Any other case is an error and must be rejected.  This includes the case of

>     the plural field being specified and the singular not.  Rationale: It's

Reviewer 2: nit: I was confused by a discussion of clearing/patching in the context of a create API call until I got to the next sentence. maybe lead with `Rationale: In an update operation, it's impossible ...`

Author: ACK

compatible structure. Eventually, a new API version, e.g. v7beta1,
will be forked and it can drop the singular field entirely.

#### Single-Dual ambiguity

Assume the user starts with:

```
kind: Frobber
height: 42
width: 3
param: "super"
```

On create we can set `params: ["super"]`.

On an unrelated POST (aka replace), an old client would send:

```
kind: Frobber
height: 3
width: 42
param: "super"
```

If we don't require new clients to use both singular and plural fields, a new
client would send:

```
kind: Frobber
height: 3
width: 42
params: ["super"]
```

That seems clear enough - we can assume `param: "super"`.

> But the old client could send a this, via patch:

Reviewer: naive non english native question `could send a this` is this valid or a typo?

Author: typo

```
PATCH  /frobbers/1
{ param: "" }
```

That gets applied to the old object before registry code can see it, and we end up with:

```
kind: Frobber
height: 42
width: 3
params: ["super"]
```

By the previous logic, we would copy `params[0]` to `param` and end up with
`param: "super"`.  But that's not what the user wanted and more importantly is
different than what happened before we pluralized.

> To totally disambiguate that, we require users of plural to always deal with
> singular, too.

Reviewer 2: nit: s/deal with/specify/, e.g.

```suggestion
To disambiguate that, we require users of the plural field to always specify the singular field as well.
```

Author: ACK
